File: java/src/main/java/Pipeline.java
import dependencies.Config;
import dependencies.Emailer;
import dependencies.Logger;
import dependencies.Project;
import io.vavr.collection.Traversable;
import io.vavr.control.Try;

import java.util.List;
import java.util.function.Function;

class TestsFailedException extends RuntimeException {
  public TestsFailedException() {
    super("Tests failed");
  }
}

class DeploymentFailedException extends RuntimeException {
  public DeploymentFailedException() {
    super("Deployment failed");
  }
}

public class Pipeline {
  private final Config config;
  private final Emailer emailer;
  private final Logger log;

  public Pipeline(Config config, Emailer emailer, Logger log) {
    this.config = config;
    this.emailer = emailer;
    this.log = log;
  }

  public void run(Project project) {
    Try.traverse(steps(this::runTests, this::runDeployment),
                 f -> f.apply(project).andThen(log::info))
       .onFailure(this::logError)
       .map(Traversable::last)
       .recover(Throwable::getMessage)
       .andThen(this::sendNotification);
  }

  @SafeVarargs
  private List<Function<Project, Try<String>>> steps(Function<Project, Try<String>>... fs) {
    return List.of(fs);
  }

  private Try<String> runTests(Project project) {
    if (!project.hasTests()) {
      return Try.success("No tests");
    }
    if (isSuccessful(project.runTests())) {
      return Try.success("Tests passed");
    }
    return Try.failure(new TestsFailedException());
  }

  private Try<String> runDeployment(Project project) {
    if (isSuccessful(project.deploy())) {
      return Try.success("Deployment successful");
    }
    return Try.failure(new DeploymentFailedException());
  }

  private void sendNotification(String message) {
    if (config.sendEmailSummary()) {
      log.info("Sending email");
      emailer.send(message);
    } else {
      log.info("Email disabled");
    }
  }

  private void logError(Throwable e) {
    log.error(e.getMessage());
  }

  private boolean isSuccessful(String s) {
    return "success".equals(s);
  }
}

File: java/src/main/kotlin/Pipeline.kt
import dependencies.Config
import dependencies.Emailer
import dependencies.Logger
import dependencies.Project
import io.vavr.control.Try

typealias PipelineStep = (Project) -> Try<String>

class Pipeline(private val config: Config,
               private val emailer: Emailer,
               private val log: Logger) {
  class TestFailedError : RuntimeException("Tests failed")
  class DeploymentFailedError : RuntimeException("Deployment failed")

  private val pipelineSteps = listOf<PipelineStep>(::runTests, ::deploy)

  fun run(project: Project) {
    Try.traverse(pipelineSteps) { runStepOn -> runStepOn(project).onSuccess(log::info) }
      .map { it.last() }
      .onFailure(::logError)
      .recover(Throwable::message)
      .flatMap(::sendEmailNotification)
      .onSuccess(log::info)
  }

  private fun logError(throwable: Throwable) = log.error(throwable.message)

  private fun runTests(project: Project): Try<String> = when {
    !project.hasTests()                -> Try.success("No tests")
    !project.runTests().isSuccessful() -> Try.failure(TestFailedError())
    else                               -> Try.success("Tests passed")
  }

  private fun deploy(project: Project): Try<String> = when {
    !project.deploy().isSuccessful() -> Try.failure(DeploymentFailedError())
    else                             -> Try.success("Deployment successful")
  }

  private fun String.isSuccessful() = "success" == this

  private fun sendEmailNotification(message: String): Try<String> {
    if (!config.sendEmailSummary()) {
      return Try.success("Email disabled")
    }
    emailer.send(message)
    return Try.success("Sending email")
  }
}